File: bluetooth/device-info-service.js
var util = require('util');
var bleno = require('bleno');

var ManufacturerCharacteristic = function(manufacturer) {
    bleno.Characteristic.call(this, {
        uuid: '2A29',
        properties: ['read'],
        descriptors: [
            new bleno.Descriptor({
               uuid: '2901',
               value: 'Manufacturer Name'
            }),
            new bleno.Descriptor({
               uuid: '2904',
               value: new Buffer([0x19, 0x00, 0x27, 0x00, 0x01, 0x00, 0x00])
            }),
        ],
        value: new Buffer(manufacturer),
    }
    )
}

util.inherits(ManufacturerCharacteristic, bleno.Characteristic);

var ModelCharacteristic = function(model) {
    bleno.Characteristic.call(this, {
        uuid: '2A24',
        properties: ['read'],
        descriptors: [
            new bleno.Descriptor({
               uuid: '2901',
               value: 'Model Number'
            }),
            new bleno.Descriptor({
               uuid: '2904',
               value: new Buffer([0x19, 0x00, 0x27, 0x00, 0x01, 0x00, 0x00])
            }),
        ],
        value: new Buffer(model),
    }
    )
}

util.inherits(ModelCharacteristic, bleno.Characteristic);

var SerialCharacteristic = function(serial) {
    bleno.Characteristic.call(this, {
        uuid: '2A25',
        properties: ['read'],
        descriptors: [
            new bleno.Descriptor({
               uuid: '2901',
               value: 'Serial Number'
            }),
            new bleno.Descriptor({
               uuid: '2904',
               value: new Buffer([0x19, 0x00, 0x27, 0x00, 0x01, 0x00, 0x00])
            }),
        ],
        value: new Buffer(serial),
    }
    )
}

util.inherits(SerialCharacteristic, bleno.Characteristic);

function DeviceInfoService(manufacturer, model, serial) {
    bleno.PrimaryService.call(this, {
        uuid: '180a',
        characteristics: [
            new ManufacturerCharacteristic(manufacturer),
            new ModelCharacteristic(model),
            new SerialCharacteristic(serial),
        ]
    });
}

util.inherits(DeviceInfoService, bleno.PrimaryService);

module.exports = DeviceInfoService;

File: bluetooth/device_info_service.py
from pybleno import *


class ManufacturerCharacteristic(Characteristic):
    def __init__(self, manufacturer):
        Characteristic.__init__(self, {
            'uuid': '2A29',
            'properties': ['read'],
            'value': bytes(manufacturer, 'utf-8'),
            'descriptors': [
                Descriptor({
                    'uuid': '2901',
                    'value': bytes('Manufacturer Name', 'utf-8')
                }),
                Descriptor({
                    'uuid': '2904',
                    'value': array.array('B', [0x19, 0x00, 0x27, 0x00, 0x01, 0x00, 0x00])
                })
            ]
        })


class ModelCharacteristic(Characteristic):
    def __init__(self, model):
        Characteristic.__init__(self, {
            'uuid': '2A24',
            'properties': ['read'],
            'value': bytes(model, 'utf-8'),
            'descriptors': [
                Descriptor({
                    'uuid': '2901',
                    'value': bytes('Model Number', 'utf-8')
                }),
                Descriptor({
                    'uuid': '2904',
                    'value': array.array('B', [0x19, 0x00, 0x27, 0x00, 0x01, 0x00, 0x00])
                })
            ]
        })


class SerialCharacteristic(Characteristic):
    def __init__(self, serial):
        Characteristic.__init__(self, {
            'uuid': '2A25',
            'properties': ['read'],
            'value': bytes(serial, 'utf-8'),
            'descriptors': [
                Descriptor({
                    'uuid': '2901',
                    'value': bytes('Serial Number', 'utf-8')
                }),
                Descriptor({
                    'uuid': '2904',
                    'value': array.array('B', [0x19, 0x00, 0x27, 0x00, 0x01, 0x00, 0x00])
                })
            ]
        })


class DeviceInfoService(BlenoPrimaryService):
    uuid = '180a'

    def __init__(self, manufacturer, model, serial):
        BlenoPrimaryService.__init__(self, {
            'uuid': self.uuid,
            'characteristics': [
                ManufacturerCharacteristic(manufacturer),
                ModelCharacteristic(model),
                SerialCharacteristic(serial)
            ]
        })
